### ISSUE TITLE
Relearn protection in 1.1.x

### ISSUE THREAD

**Issue by moisseev** [vstakhov/rspamd#490]
Since I've upgraded to 1.1 I 'm getting `HTTP error: 404, <some MID> has been already learned as ham, ignore it.` error for some messages that I quite sure haven't learned before.
What's supposed to be stored in `learn_cache.sqlite`? Hashes of whole messages or just of some parts of them?

**Comment by mnik247** [comment]
Confirn too.
After upgrade from 1.0.11 to 1.1.0 (centos 6 x64) I get strange learning spam/ham with error:
1) first:
Results for file: /tmp/spam-20160119-220021/1525b790a90-12 (0.297 seconds)
IO read error: Connection reset by peer
Results for file: /tmp/spam-20160119-220021/1525b790a90-9 (0.286 seconds)
IO read error: Connection reset by peer
Results for file: /tmp/spam-20160119-220021/1525b790a90-8 (0.284 seconds)
IO read error: Connection reset by peer
Results for file: /tmp/spam-20160119-220021/1525b790a90-10 (0.281 seconds)
IO read error: unexpected EOF
Results for file: /tmp/spam-20160119-220021/1525b790a90-13 (0.271 seconds)
IO read error: Connection reset by peer
Results for file: /tmp/spam-20160119-220021/1525b790a90-15 (0.270 seconds)
IO read error: Connection reset by peer
Results for file: /tmp/spam-20160119-220021/1525b790a90-0 (0.268 seconds)
IO read error: Connection reset by peer
Results for file: /tmp/spam-20160119-220021/1525b790a90-4 (0.265 seconds)
IO read error: Connection reset by peer
Results for file: /tmp/spam-20160119-220021/1525b790a90-6 (2.139 seconds)
IO read error: Connection reset by peer
Results for file: /tmp/spam-20160119-220021/1525b790a90-17 (2.096 seconds)
IO read error: unexpected EOF
Results for file: /tmp/spam-20160119-220021/1525b790a90-16 (2.087 seconds)
IO read error: Connection reset by peer
Results for file: /tmp/spam-20160119-220021/1525b790a90-7 (2.078 seconds)
IO read error: Connection reset by peer
Results for file: /tmp/spam-20160119-220021/1525b790a90-11 (4.125 seconds)
IO read error: Connection reset by peer
Results for file: /tmp/spam-20160119-220021/1525b790a90-2 (4.123 seconds)
IO read error: Connection reset by peer
Results for file: /tmp/spam-20160119-220021/1525b790a90-3 (4.121 seconds)
IO read error: unexpected EOF
Results for file: /tmp/spam-20160119-220021/1525b790a90-14 (4.115 seconds)
IO read error: Connection reset by peer
Results for file: /tmp/spam-20160119-220021/1525b790a90-5 (2.022 seconds)
HTTP error: 404, <E1aLWDs-0007tM-Qw@st12.mi6.kiev.ua> has been already learned as spam, ignore it
Results for file: /tmp/spam-20160119-220021/1525b790a90-1 (2.222 seconds)
IO read error: unexpected EOF

2) second:
Results for file: /tmp/spam-20160119-220021/1525b790a90-10 (0.033 seconds)
HTTP error: 404, <E1aLUed-0006OA-Ro@peper.pp.ua> has been already learned as spam, ignore it
Results for file: /tmp/spam-20160119-220021/1525b790a90-13 (0.033 seconds)
HTTP error: 404, <7955BF176DD3D85C55CB22C0192F83C0@MXS.MAIL.RU> has been already learned as spam, ignore it
Results for file: /tmp/spam-20160119-220021/1525b790a90-12 (0.057 seconds)
HTTP error: 404, <5257E1B2B96D43904C4900BD490583A8@bizneroa.co.ua> has been already learned as spam, ignore it
Results for file: /tmp/spam-20160119-220021/1525b790a90-9 (0.075 seconds)
HTTP error: 404, <4346F0A3A87C52815D5811AC5838DF0F@bizneroa.co.ua> has been already learned as spam, ignore it
Results for file: /tmp/spam-20160119-220021/1525b790a90-8 (0.094 seconds)
HTTP error: 404, <C5C076252EFAD407DBDE972ADE957695@bizneroa.co.ua> has been already learned as spam, ignore it
Results for file: /tmp/spam-20160119-220021/1525b790a90-15 (0.152 seconds)
HTTP error: 404, <20DA931C3CE8C615C9CC8538CC75B18C@bizneroa.co.ua> has been already learned as spam, ignore it
Results for file: /tmp/spam-20160119-220021/1525b790a90-0 (0.151 seconds)
HTTP error: 404, <333680D3D80C22F12D2861DC2829EE2B@bizneroa.co.ua> has been already learned as spam, ignore it
Results for file: /tmp/spam-20160119-220021/1525b790a90-4 (0.151 seconds)
HTTP error: 404, <CCC97F2C27F3DD0ED2D79E23D7F7D09D@bizneroa.co.ua> has been already learned as spam, ignore it
Results for file: /tmp/spam-20160119-220021/1525b790a90-17 (0.012 seconds)
HTTP error: 404, <EXI405RZ-E6LV-WSOF-UDJK-XJGSMM170QG3@yandex.kz> has been already learned as spam, ignore it
Results for file: /tmp/spam-20160119-220021/1525b790a90-16 (0.012 seconds)
HTTP error: 400, <1453175219.802083875@f221.i.mail.ru> contains less tokens than required for bayes classifier: 10 < 0
Results for file: /tmp/spam-20160119-220021/1525b790a90-6 (0.151 seconds)
IO read error: Connection reset by peer
Results for file: /tmp/spam-20160119-220021/1525b790a90-7 (0.150 seconds)
IO read error: Connection reset by peer
Results for file: /tmp/spam-20160119-220021/1525b790a90-11 (0.150 seconds)
IO read error: Connection reset by peer
Results for file: /tmp/spam-20160119-220021/1525b790a90-2 (0.150 seconds)
IO read error: Connection reset by peer
Results for file: /tmp/spam-20160119-220021/1525b790a90-3 (0.149 seconds)
IO read error: unexpected EOF
Results for file: /tmp/spam-20160119-220021/1525b790a90-14 (0.149 seconds)
IO read error: unexpected EOF
Results for file: /tmp/spam-20160119-220021/1525b790a90-5 (2.023 seconds)
HTTP error: 404, <E1aLWDs-0007tM-Qw@st12.mi6.kiev.ua> has been already learned as spam, ignore it
Results for file: /tmp/spam-20160119-220021/1525b790a90-1 (2.052 seconds)
HTTP error: 404, <B9BC0A595286A87BA7A2EB56A2689C9B@bizneroa.co.ua> has been already learned as spam, ignore it

3) again:
Results for file: /tmp/spam-20160119-220021/1525b790a90-10 (0.011 seconds)
HTTP error: 404, <E1aLUed-0006OA-Ro@peper.pp.ua> has been already learned as spam, ignore it
Results for file: /tmp/spam-20160119-220021/1525b790a90-13 (0.030 seconds)
HTTP error: 404, <7955BF176DD3D85C55CB22C0192F83C0@MXS.MAIL.RU> has been already learned as spam, ignore it
Results for file: /tmp/spam-20160119-220021/1525b790a90-15 (0.088 seconds)
HTTP error: 404, <20DA931C3CE8C615C9CC8538CC75B18C@bizneroa.co.ua> has been already learned as spam, ignore it
Results for file: /tmp/spam-20160119-220021/1525b790a90-0 (0.087 seconds)
HTTP error: 404, <333680D3D80C22F12D2861DC2829EE2B@bizneroa.co.ua> has been already learned as spam, ignore it
Results for file: /tmp/spam-20160119-220021/1525b790a90-4 (0.087 seconds)
HTTP error: 404, <CCC97F2C27F3DD0ED2D79E23D7F7D09D@bizneroa.co.ua> has been already learned as spam, ignore it
Results for file: /tmp/spam-20160119-220021/1525b790a90-12 (0.147 seconds)
HTTP error: 404, <5257E1B2B96D43904C4900BD490583A8@bizneroa.co.ua> has been already learned as spam, ignore it
Results for file: /tmp/spam-20160119-220021/1525b790a90-9 (0.147 seconds)
HTTP error: 404, <4346F0A3A87C52815D5811AC5838DF0F@bizneroa.co.ua> has been already learned as spam, ignore it
Results for file: /tmp/spam-20160119-220021/1525b790a90-8 (0.147 seconds)
HTTP error: 404, <C5C076252EFAD407DBDE972ADE957695@bizneroa.co.ua> has been already learned as spam, ignore it
Results for file: /tmp/spam-20160119-220021/1525b790a90-17 (0.012 seconds)
HTTP error: 404, <EXI405RZ-E6LV-WSOF-UDJK-XJGSMM170QG3@yandex.kz> has been already learned as spam, ignore it
Results for file: /tmp/spam-20160119-220021/1525b790a90-16 (0.012 seconds)
HTTP error: 400, <1453175219.802083875@f221.i.mail.ru> contains less tokens than required for bayes classifier: 10 < 0
Results for file: /tmp/spam-20160119-220021/1525b790a90-3 (0.019 seconds)
HTTP error: 404, <E1aLWJR-0002M2-Tu@st15.mi6.kiev.ua> has been already learned as spam, ignore it
Results for file: /tmp/spam-20160119-220021/1525b790a90-14 (0.019 seconds)
HTTP error: 404, <7F85CC4026F2DC6FB3B6FF41352422BE@79.104.41.158> has been already learned as spam, ignore it
Results for file: /tmp/spam-20160119-220021/1525b790a90-6 (0.047 seconds)
HTTP error: 404, <03ea01d152d7$510d2fc0$a9fbe796@izyuprs> has been already learned as spam, ignore it
Results for file: /tmp/spam-20160119-220021/1525b790a90-7 (0.102 seconds)
HTTP error: 404, <4540F6A5AE7A54875B5E17AA5E4EB29A@bizneroa.co.ua> has been already learned as spam, ignore it
Results for file: /tmp/spam-20160119-220021/1525b790a90-11 (0.102 seconds)
HTTP error: 404, <A4A117444F9BB566BABFF64BBFB91B16@bizneroa.co.ua> has been already learned as spam, ignore it
Results for file: /tmp/spam-20160119-220021/1525b790a90-2 (0.102 seconds)
HTTP error: 404, <8B8E386B60B49A499590D96490FAB2B6@bizneroa.co.ua> has been already learned as spam, ignore it
Results for file: /tmp/spam-20160119-220021/1525b790a90-5 (0.003 seconds)
HTTP error: 404, <E1aLWDs-0007tM-Qw@st12.mi6.kiev.ua> has been already learned as spam, ignore it
Results for file: /tmp/spam-20160119-220021/1525b790a90-1 (0.022 seconds)
HTTP error: 404, <B9BC0A595286A87BA7A2EB56A2689C9B@bizneroa.co.ua> has been already learned as spam, ignore it

I don't see this errors before upgrade to 1.1.0 .

**Comment by vstakhov** [comment]
Relearn protection was broken in 1.0. And I suppose there are some issues for sqlite backend. I need to check one thing...

**Comment by vstakhov** [comment]
Well, I've added some fix to the master that *might* help with this issue. However, it makes all previous data in learn cache useless :(

**Comment by vstakhov** [comment]
EOF behaviour should be studied more carefully. Do you see anything in rspamd logs?

**Comment by moisseev** [comment]
387a965 has fixed the relearn protection. I can't say anything about EOF, I didn't have that issue.

**Comment by mnik247** [comment]
@vstakhov   
1) 2016-01-20 00:47:22 #27177(controller) <1c3e1d>; task; rspamd_message_parse: loaded message; id: <7F85CC4026F2DC6FB3B6FF41352422BE@79.104.41.158>; queue-id:
<undef>
2016-01-20 00:47:22 #2673(main) <328461>; main; rspamd_cld_handler: controller process 27177 terminated abnormally with exit code 127
2) 2016-01-20 00:44:43 #3504(controller) <cc9a22>; task; rspamd_message_parse: loaded message; id: <7955BF176DD3D85C55CB22C0192F83C0@MXS.MAIL.RU>; queue-id: <un
def>
2016-01-20 00:44:43 #2673(main) <328461>; main; rspamd_cld_handler: controller process 3504 terminated abnormally with exit code 127

**Comment by vstakhov** [comment]
Any chance to get core file?

**Comment by mnik247** [comment]
@vstakhov  
What do you mean "get core file"?

**Comment by vstakhov** [comment]
It seems that your controller has been crashed. So it should have dropped a core file in this case.

**Comment by moisseev** [comment]
A file named `programname.core` in the process's working directory.
See `man 5 core` for details.

**Comment by vstakhov** [comment]
@moisseev, well my question is slightly different. I did the following fix: I've changed words hash to stats token hash to be stored in the statistics cache. Words hash was much slower and less reliable. However, it was broken in 1.0 completely - it didn't prevent from relearning. So my question now is whether to keep the old hash and check it for some time (would give some extra CPU consumption) or just throw it completely. What are your thoughts?

**Comment by moisseev** [comment]
It doesn't matter for me personally. I'm using rspamd on very small mail servers, so I have no issues with performance. On the other hand, I do manual learning only, not automatic. So, I can live without relearning protection too. You should better ask users who has huge servers with really big traffic.

I believe probability of relearning the same message should decrease over time. May be you should drop old hash support, but warn users about that. So they would stop learning for couple of days after update?

**Comment by mnik247** [comment]
@vstakhov
I can't get rspamd core dump.
I set all setting for core dump, get test core dump, but not rspamd. 
May be I should reboot, but I can't now.
But I finally fixed errors with rspamc learning - I updated sqlite: sqlite-3.6.20 -> sqlite-3.7.17.
And after update sqlite not present other errors in rspams.log like:
main; rspamd_cld_handler: fuzzy process 7756 terminated abnormally with exit code 127
main; rspamd_cld_handler: controller process 7770 terminated abnormally with exit code 127
